fix: update log level handling

catsby opened the pull request:
## Short Description

This PR updates Maru's logging to match the standard library's `log/slog` package, and appropriately filters log messages according to levels. Tests included and feedback very welcome.

## Long Description

The current implementation of log level handling is misleading because our [custom handler](https://github.com/defenseunicorns/maru-runner/blob/535258256cd1714f55e643075ae15919316e4aec/src/message/slog.go#L20-L23) doesn't filter by level. The code comment mentions logging functions already being aware if they are allowed to be called, but in practice I have not found that to be true. This is further complicated by Maru's log levels not matching what `log/slog` uses both in values and in ordering. 

Example: slog uses these levels/ordering

```go
const (
	LevelDebug Level = -4
	LevelInfo  Level = 0
	LevelWarn  Level = 4
	LevelError Level = 8
)
```

(See here for source and long explaination to the numbers used: https://pkg.go.dev/log/slog#Level)

Where as Maru uses these levels/ordering:
```go
const (
	// WarnLevel level. Non-critical entries that deserve eyes.
	WarnLevel LogLevel = iota
	// InfoLevel level. General operational entries about what's going on inside the
	// application.
	InfoLevel
	// DebugLevel level. Usually only enabled when debugging. Very verbose logging.
	DebugLevel
	// TraceLevel level. Designates finer-grained informational events than the Debug.
	TraceLevel
)
```

(source: [src/message/logging.go#L16-L29](https://github.com/defenseunicorns/maru-runner/blob/535258256cd1714f55e643075ae15919316e4aec/src/message/logging.go#L16-L29))

In theory, implementing the [`Handler` interface](https://cs.opensource.google/go/go/+/refs/tags/go1.23.1:src/log/slog/handler.go;l=29-33) would have us ignore messages with a lower level than the one set. Maru sets it's default log level to `InfoLevel` (from Maru's constants, not slog's) which results in a value of `1`. However because our `InfoLevel` is `1` and slog's `LevelInfo` is `0`, then in theory Maru's logger should actually ignore it because it's a lower level. A call to `message.SLog.Info()` ([ex. in utils.go](https://github.com/defenseunicorns/maru-runner/blob/535258256cd1714f55e643075ae15919316e4aec/src/pkg/utils/utils.go#L35)) invokes slog's `Info()` method, which calls an internal [method `log()`](https://cs.opensource.google/go/go/+/refs/tags/go1.23.1:src/log/slog/logger.go;l=208-211) with slog's `LevelInfo` level which is a value of `0`. The [`log()` method](https://cs.opensource.google/go/go/+/refs/tags/go1.23.1:src/log/slog/logger.go;l=238-258) calls the handlers `Enabled()` method to see if it should log the message, which for maru always returns `true`, and then passes it on to the handlers `Handle()` method (Maru's `Handle()` method [here](https://github.com/defenseunicorns/maru-runner/blob/535258256cd1714f55e643075ae15919316e4aec/src/message/slog.go#L37)) and we print the message, even though we shouldn't because it's a lower value than our default.   

The net effect is setting the log level with `--log-level` has no effect, except for when setting to `trace`, in which case we output pterm debugging information with lines for errors/traces and debug calls. Otherwise Maru will always output all of the logs from `Debug()`, `Info()`, `Warn()` or `Error()` regardless of the level set. 

This PR updates Maru's log level constants both in values and ordering to match the standard library `log/slog` package, and implements the `Enabled()` method in the handler to filter correctly based on level. The net effect of this PR should be benign in that we aren't changing any usage details, but going forward we'll correctly output logs according to the set level. 

**NOTE:** This PR is based off of [my other PR (#140) for changing the `Makefile`](https://github.com/defenseunicorns/maru-runner/pull/140) with regards to `test-unit` and `test-e2e` tests. 

## Test output:

```zsh
 ➜ go test ./src/message/... -v -count=1
=== RUN   Test_LogLevel_Diff
=== RUN   Test_LogLevel_Diff/DebugLevel
=== RUN   Test_LogLevel_Diff/InfoLevel
=== RUN   Test_LogLevel_Diff/InfoWarnLevel
=== RUN   Test_LogLevel_Diff/WarnInfoLevel
=== RUN   Test_LogLevel_Diff/InfoTraceLevel
=== RUN   Test_LogLevel_Diff/TraceInfoLevel
=== RUN   Test_LogLevel_Diff/TraceLevel
--- PASS: Test_LogLevel_Diff (0.00s)
    --- PASS: Test_LogLevel_Diff/DebugLevel (0.00s)
    --- PASS: Test_LogLevel_Diff/InfoLevel (0.00s)
    --- PASS: Test_LogLevel_Diff/InfoWarnLevel (0.00s)
    --- PASS: Test_LogLevel_Diff/WarnInfoLevel (0.00s)
    --- PASS: Test_LogLevel_Diff/InfoTraceLevel (0.00s)
    --- PASS: Test_LogLevel_Diff/TraceInfoLevel (0.00s)
    --- PASS: Test_LogLevel_Diff/TraceLevel (0.00s)
PASS
ok      github.com/defenseunicorns/maru-runner/src/message      0.235s
```

## Related

- https://github.com/defenseunicorns/uds-cli/issues/916

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Other (security config, docs update, etc)

## Checklist before merging

- [x] Test, docs, adr added or updated as needed
- [x] [Contributor Guide Steps](https://github.com/defenseunicorns/maru-runner/blob/main/CONTRIBUTING.md) followed

Racer159 commented on src/message/logging.go:
```diff
@@ -17,18 +17,23 @@ import (
+	WarnLevel LogLevel = 4
 	// TraceLevel level. Designates finer-grained informational events than the Debug.
-	TraceLevel
+	TraceLevel LogLevel = 8
```
Trace level was meant to be a level below debug level (with more output in the logs) - Maru is a bit weird bc it was a bunch of refactored code that wound up being smushed quickly into a repo. The difference you see at Trace is due to this: https://github.com/defenseunicorns/maru-runner/blob/77f52c9d13281e0e0d9b843ccc72ad8727f6789d/src/message/message.go#L70 which should probably be refactored to use slogs levels with TraceLevel likely being a `-8` that is checked in those printers.

We should probably also add an additional errorlevel to properly handle the error messaging if that level is chosen.

catsby replied:
Updated!

Racer159 commented on src/message/logging.go:
```diff
@@ -59,7 +64,8 @@ func LogFileLocation() string {
 	logLevel = lvl
-	if logLevel >= DebugLevel {
+	// Enable pterm debug messages if the log level is Trace or Debug
+	if logLevel == DebugLevel || logLevel == TraceLevel {
```
This would now be <= Debug

catsby replied:
Done

Racer159 commented on src/message/message.go:
```diff
@@ -67,7 +67,8 @@ func paragraph(format string, a ...any) string {
 
 func debugPrinter(offset int, a ...any) {
-	printer := pterm.Debug.WithShowLineNumber(logLevel > 2).WithLineNumberOffset(offset)
+	showLines := logLevel == DebugLevel || logLevel == TraceLevel
```
lines only showed on TraceLevel before so this should probably be <= Trace

catsby replied:
Done

Racer159 commented on src/message/slog.go:
```diff
@@ -17,9 +17,20 @@ var (
+//
+// Examples:
+//
+//	SetLogLevel(DebugLevel) // show everything
```
```suggestion
//	SetLogLevel(TraceLevel) // show everything
```

catsby replied:
I did make this change but didn't notice this was a commit suggestion, so I'm going close it but that change is in there:

```go
//	SetLogLevel(TraceLevel) // show everything, with file names and line numbers
//	SetLogLevel(DebugLevel) // show everything
//	SetLogLevel(InfoLevel)  // show info and above (does not show debug logs)
//	SetLogLevel(WarnLevel)  // show warn and above (does not show debug/info logs)
//	SetLogLevel(ErrorLevel)  // show only errors (does not show debug/info/warn logs)
```

catsby commented:
@Racer159 thanks again for the review. I applied your feedback and added an `ErrorLevel` level. I also updated the tests to check for the extra lines of output when using `TraceLevel`. In hindsight the way I went about checking feels clunky, so if you have ideas on improving it I'm all for it.

Racer159 commented on src/message/message.go:
```diff
@@ -67,7 +67,7 @@ func paragraph(format string, a ...any) string {
+	printer := pterm.Debug.WithShowLineNumber(logLevel <= TraceLevel).WithLineNumberOffset(offset)
 	now := time.Now().Format(time.RFC3339)
 	// prepend to a
 	a = append([]any{now, " - "}, a...)
```
pterm.error down on `89` should also be logLevel <= TraceLevel

catsby replied:
👍 I didn't see that, thanks. Updated in de6040d
